Issue 474 distrib flattening

luciansmith opened the pull request:
This addresses https://sourceforge.net/p/sbml/libsbml/474/?limit=25 and implements and tests distrib/comp flattening.

Something is still wonky with the flattening converter, because I could not get things to work as expected when 'distrib' was *not* enabled in the libsbml build:  it ends up coming through as an unknown package, but telling the converter to strip unflattenable packages apparently doesn't include unknown packages.  But that aside, everything else is now working and tested.

There were also two flattening tests that were failing that this commit fixes.  This means that travis isn't testing comp flattening, and may not be testing the packages at all?  That should also be fixed, but not here.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [ ] I have updated all documentation necessary.
- [ ] I have checked spelling in (new) comments.

## Testing
- [X] Testing is done automatically and codecov shows test coverage
- [ ] This cannot be tested automatically <!-- describe how it has been tested -->

luciansmith committed:
Implement distrib flattening.

These changes should be all that are needed to flatten 'distrib' models.  Still need to add tests, but the basics are here.

Also fixed up a couple things in the base flattening algorithms, regarding 'set/getId' vs 'set/getIdAttribute'.

luciansmith committed:
Fix flattening of priorities.

This fix probably goes along with a different recent fix where mathml that needed to be converted with a conversion factor where the root ASTNode needed to be converted was not being handled properly.  This means that this test was ensuring that we did things wrong: now that that was fixed, this version is actually correct.

luciansmith committed:
Add tests for distrib flattening.

Also revamp package flattening tests in general so everything uses the same base function instead of copy/pasted code everywhere.

Also, one of the tests would have never worked, since it had the wrong filename to compare it to ('conversion_factor2_param_flat').

skeating reviewed (changes requested):
This does not pass all tests in test_comp_extensions.

Also, the issues you highlight would suggest more time/thought/investigation is needed

luciansmith committed:
Fix another test that was broken by a different earlier fix.

This is another test model that is now correctly flattened, but was incorrectly flattened in the past, since the priorities didn't properly use the conversion properties.

Doesn't have anything to do with distrib flattening, but worth fixing.

luciansmith commented:
The failing test is now fixed, though it was failing in development as well.

The issues I brought up should be addressed, but apply to *all* package flattening, not just distrib.  This PR brings distrib up to par with all the other packages, which I think is worth doing.
